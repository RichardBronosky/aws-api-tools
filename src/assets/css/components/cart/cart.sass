@import '../../mixins'
@import '../../variables'

.cart-page
  margin-top: 1em
  .bold
    font-weight: 600
  h1
    color: $dark-grey
    line-height: 1
    font-size: $txt-xlarge
    font-weight: 300
    padding: 0
    text-align: left
    margin-bottom: 1.2rem
    margin-top: 1.5rem
    +medium-down
      text-align: center
  .cart-product-list-container
    margin-right: 1em
    +medium-down
      margin-top: 0
    .financing-disclaimer
      color: $red
      text-transform: uppercase
    .slider
      margin-top: 1.5em
      +large-up
        margin-top: 0
      h2 
        font-size: $txt-medium
      .slick-slider
        +large-up
          margin-left: -2em
          margin-right: -2em
        .buy-now-popup
          .buy-now
            font-size: $txt-xsmall
  
  .checkout-sticky
    .order-detail-expander
      padding: 0

  .checkout-container
    position: fixed
    bottom: 0
    width: 100%!important
    z-index: 1
    border-top: 2px solid $grey
    margin-left: 0 !important
    +large-up
      margin-top: 0 !important
      position: -webkit-sticky
      position: sticky
      top: 10.375em!important
      right: 0
      height: 100%
      width: 23%!important
      z-index: 8
      border: 0
    .checkout-sticky
      margin-top: 0
.cart-bottom-checkout
  margin: 1rem 0 .2rem 0
  background-color: $primary-invert
  .checkout-sticky
    .checkout-info
      &.bottom
        margin: 0
  .button-container
    display: flex
    .checkout-btn-container
      width: 100% !important
    .paypal-btn-container
      margin-top: .5rem
      padding: 0
      +medium-up
        display: none!important
      .paypal
        padding: 0 1rem
        .loader
          height: 2rem
          width: 100%
          +medium-down
            width: 33.75%
