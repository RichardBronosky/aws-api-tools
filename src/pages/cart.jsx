import React from 'react'
import Helmet from 'react-helmet'
import Layout from '../components/layout'
import Cart from '../components/cart'

export default ({ props }) => (
  <Layout { ...props } cartPage>
    <Helmet title={ 'Shopping Cart - Rooms To Go' } meta={ [{ name: 'robots', content: 'noindex, nofollow' }] } />
    <Cart />
  </Layout>
)
