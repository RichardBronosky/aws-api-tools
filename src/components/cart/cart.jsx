import React from 'react'
import classNames from 'classnames'
import CondensedContact from '../shared/condensed-contact-links'
import CartProduct from './cart-parts/cart-product'
import CreditCardBanner from './cart-parts/credit-card-banner'
import CartEmpty from './cart-parts/cart-empty'
import CheckoutSticky from '../checkout/checkout-sticky/checkout-sticky'
import PaymentMethods from './cart-parts/payment-methods'
import PromotionContentGroup from '../shared/promotion-content-group'
import { removeUnavailableItems } from '../../lib/helpers/cart'
import '../../assets/css/components/cart/cart.sass'

export default ({
  cart,
  isMobile,
  onAddActiveAddon,
  onRemoveActiveAddon,
  skusNotAvailable,
  discount,
  showPayPal,
  promoTargetSkus,
  setCartState,
}) => (
  <div className="cart-page">
    <h1 id="main-cart-title" tabIndex="-1">
      CART
    </h1>
    <div
      className={ classNames('grid-x', {
        'grid-margin-x': isMobile,
      }) }
    >
      <div
        className={ classNames('cart-product-list-container small-12 medium-12 large-9 grid-x grid-padding-y', {
          cell: isMobile,
        }) }
      >
        { cart &&
          cart.cartItems &&
          cart.cartItems.map((item, index) => (
            <CartProduct
              key={ index }
              product={ item.product }
              index={ index }
              productCount={ cart.cartItems.length }
              quantity={ item.quantity }
              price={ item.price }
              unavailableItem={ skusNotAvailable.filter(prod => prod.sku === item.sku)[0] }
              activeAddons={ item.activeAddons }
              onAddActiveAddon={ onAddActiveAddon }
              onRemoveActiveAddon={ onRemoveActiveAddon }
            />
          )) }
        { (!cart || (cart && cart.cartItems && cart.cartItems.length < 1) || (cart && !cart.cartItems)) && <CartEmpty /> }
        { cart && cart.cartItems && cart.cartItems.length > 0 && (
          <>
            <PromotionContentGroup targetSkus={ promoTargetSkus } isCart />
            <div className="cart-bottom-checkout cell small-12 grid-x">
              <div className="cell large-8 payment-methods">
                <PaymentMethods />
              </div>
              <CheckoutSticky
                cart={ cart }
                productsAvailable={ skusNotAvailable.length < 1 }
                removeUnavailableItems={ () => removeUnavailableItems(setCartState, skusNotAvailable) }
                onlyButtons
                discount={ discount }
                showPayPal={ showPayPal }
              />
            </div>
          </>
        ) }
        { cart && cart.cartItems && cart.cartItems.length > 0 && (
          <>
            <p className="financing-disclaimer cell small-12">*when paying with a rooms to go credit card</p>
            <CreditCardBanner />
          </>
        ) }
        <CondensedContact />
      </div>
      <div className="checkout-container cell small-12 large-2 grid-margin-y">
        { cart && cart.cartItems && cart.cartItems.length > 0 && (
          <CheckoutSticky
            cart={ cart }
            isMobile={ isMobile }
            productsAvailable={ skusNotAvailable.length < 1 }
            removeUnavailableItems={ () => removeUnavailableItems(setCartState, skusNotAvailable) }
            rightSticky
            discount={ discount }
            showPayPal={ showPayPal }
          />
        ) }
      </div>
    </div>
  </div>
)
