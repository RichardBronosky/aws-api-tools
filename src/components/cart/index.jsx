import React from 'react'
import { connect } from 'react-redux'
import { getLastParamOfPath } from '../../lib/helpers/string-helper'
import { verifyAndUpdateCart, cartUpdate, getPromoTargetSkus, loadCartFromOrder } from '../../lib/helpers/cart'
import { setupAnalytics } from '../../lib/helpers/google-tag-manager'
import Cart from './cart'

let apiCalled = false

class CartWrapper extends React.Component {
  state = {
    skusNotAvailable: [],
    discount: null,
    showPayPal: false,
  }

  setCartState = state => this.setState({ ...this.state, ...state })

  setApiCalled = called => {
    apiCalled = called
  }

  componentDidMount() {
    const { cart } = this.props
    const { discount, skusNotAvailable } = this.state
    setupAnalytics({ pageData: { type: 'cart', title: 'Cart', path: '/cart' } })
    const lastParam = getLastParamOfPath()
    const pattern = new RegExp('[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}')
    if (lastParam && pattern.test(lastParam)) {
      loadCartFromOrder(lastParam)
    }
    if (JSON.stringify(cart) !== JSON.stringify({ cartItems: [] })) {
      verifyAndUpdateCart(cart)
      cartUpdate(cart, skusNotAvailable, apiCalled, discount, this.setCartState, this.setApiCalled, false)
    }
  }

  componentDidUpdate(prevProps) {
    const { cart } = this.props
    const { discount, skusNotAvailable } = this.state
    if (prevProps.cart !== cart) {
      verifyAndUpdateCart(cart)
      cartUpdate(cart, skusNotAvailable, apiCalled, discount, this.setCartState, this.setApiCalled, discount !== null)
    }
  }

  render() {
    const promoTargetSkus = getPromoTargetSkus()
    return <Cart { ...this.props } { ...this.state } promoTargetSkus={ promoTargetSkus } setCartState={ this.setCartState } />
  }
}

const mapStateToProps = state => {
  return { ...state.global, ...state.cart }
}

const mapDispatchToProps = dispatch => {
  return {
    onAddActiveAddon: (productSku, addon, cartIndex) => dispatch(addActiveAddon(productSku, addon, cartIndex)),
    onRemoveActiveAddon: (productSku, addonSku, cartIndex) =>
      dispatch(removeActiveAddon(productSku, addonSku, cartIndex)),
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CartWrapper)
