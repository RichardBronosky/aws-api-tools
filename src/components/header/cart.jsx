import React from 'react'
import { connect } from 'react-redux'
import { getCartTotal } from '../../lib/helpers/cart'
import { currencyFormatUS } from '../../lib/helpers/string-helper'
import { isProductStrikeThrough } from '../../lib/helpers/product'
import RTGLink from '../shared/link'
import RemoveFromCart from '../cart/cart-parts/remove-from-cart'
import { productUrl } from '../../lib/helpers/route'

class CartPopout extends React.Component {
  render() {
    const cart = this.props.cart
    return (
      <>
        <div className="section">
          <div id="cartSubtotal" className="subtotal">
            Subtotal: { currencyFormatUS(getCartTotal(cart)) }
          </div>
        </div>
        { (!cart || (cart && typeof cart.cartItems === 'undefined')) && (
          <>
            <div className="section">
              <div className="message">Your shopping cart is empty</div>
            </div>
            { /* <div className="section">
              <div className="notice">
                <p>
                  If you saved items on a previous visit,{ ' ' }
                  <a href="/">Sign In</a> to access your saved cart.
                  </p>
              </div>
            </div> */ }
          </>
        ) }
        <div className="section sticky">
          <RTGLink
            data={ {
              slug: '/checkout',
              title: 'Checkout',
              category: 'header',
              action: 'cart interaction',
              label: 'checkout',
            } }
          >
            <span className="blue-action-btn checkout">Checkout</span>
          </RTGLink>
          <RTGLink
            data={ {
              slug: '/cart',
              title: 'View Full Cart',
              category: 'header',
              action: 'cart interaction',
              label: 'view full cart',
            } }
            className="bold-arrow-link"
          >
            View Full Cart >
          </RTGLink>
        </div>
        { cart && cart.cartItems && cart.cartItems.length > 0 && (
          <>
            { cart.cartItems &&
              cart.cartItems.map((item, index) => {
                const product = item.product
                const strikethrough = isProductStrikeThrough(product)
                return (
                  <div key={ index } className="line-item grid-x">
                    <div className="product cell small-10">
                      <RTGLink
                        data={ {
                          url: productUrl(product.title, product.sku),
                        } }
                        aria-hidden="true"
                        tabIndex="-1"
                        disabled={ product.category === 'gift-card' }
                      >
                        <div className="prod-img">
                          <img src={ product.primary_image } alt="" role="presentation" />
                        </div>
                      </RTGLink>
                      <div className="prod-details">
                        <RTGLink
                          data={ {
                            url: productUrl(product.title, product.sku),
                          } }
                          aria-describedby={ `mcq_${ index }mcp_${ index }` }
                          disabled={ product.category === 'gift-card' }
                        >
                          <div className="title">{ product.title }</div>
                        </RTGLink>
                        <div id={ 'mcq_' + index } className="qty">
                          Qty: { item.quantity }
                        </div>
                        <div id={ 'mcp_' + index } className="price">
                          { strikethrough && (
                            <p className="strikethrough">{ currencyFormatUS(item.product.list_price) }</p>
                          ) }
                          <p className={ strikethrough ? 'strikethrough-sale' : '' }>{ currencyFormatUS(item.price) }</p>
                        </div>
                      </div>
                    </div>

                    <div className="cell small-2">
                      <RemoveFromCart headerCart product={ product } index={ index } />
                    </div>
                  </div>
                )
              }) }
          </>
        ) }
      </>
    )
  }
}

const mapStateToProps = state => {
  return { ...state.cart }
}

export default connect(
  mapStateToProps,
  null
)(CartPopout)
