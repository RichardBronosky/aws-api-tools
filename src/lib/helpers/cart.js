import { fetchProductBySku, fetchProductWarehouseAvailability } from '../services/product'
import { productPrice, productOnSale, availabilityStockMessage, fetchProductOrGC } from './product'
import { store } from '../../redux/store'
import { setOrder, setCheckoutStep } from '../../redux/modules/checkout'
import {
  addProductToCart,
  updateCartProductQuantity,
  removeProductFromCart,
  setCart,
  addActiveAddon,
  removeActiveAddon as removeActiveAddonRedux,
} from '../../redux/modules/cart'
import { analyticsProduct, addToDataLayer } from './google-tag-manager'
import { saveLocalStorage, getFromBrowserStorage } from './storage'
import { taskDone, announce } from './aria-announce'
import { clearCheckoutState, getOrder, getLineItems } from './checkout/global'
import { getRegionZone, getCurrentLocation } from './geo-location'
import { createOrder, getOrder as fetchOrder, updateLineItems } from '../services/checkout'

export const getCart = () => {
  return store.getState().cart.cart
}

export const fetchAndAddActiveAddon = (product, cartIndex, addon) => {
  fetchProductBySku(addon.sku).then(fetchedAddon =>
    store.dispatch(addActiveAddon(product.sku, { ...fetchedAddon, quantity: addon.quantity }, cartIndex))
  )
}

export const removeActiveAddon = (product, addonSku, cartIndex) => {
  store.dispatch(removeActiveAddonRedux(product.sku, addonSku, cartIndex))
}

export const getLineItemQuantity = () => {
  let count = 0
  let count2 = 0
  const cart = getCart()
  if (cart && cart.cartItems && cart.cartItems.length > 0) {
    for (let i = 0, n = cart.cartItems.length; i < n; i++) {
      count += parseInt(cart.cartItems[i].quantity)
      if (cart.cartItems[i].activeAddons) {
        for (let x = 0, y = cart.cartItems[i].activeAddons.length; x < y; x++) {
          count += parseInt(cart.cartItems[i].quantity * cart.cartItems[i].activeAddons[x].quantity)
        }
      }
      if (cart.cartItems[i].product && cart.cartItems[i].product.addon_items) {
        const requiredAddons = cart.cartItems[i].product.addon_items.filter(addon => addon.addon_required)
        if (requiredAddons.length > 0) {
          for (let x = 0, y = requiredAddons.length; x < y; x++) {
            if (
              !cart.cartItems[i].activeAddons ||
              (cart.cartItems[i].activeAddons &&
                cart.cartItems[i].activeAddons.filter(addon => addon.sku === requiredAddons[x].sku).length < 1)
            ) {
              count2 += cart.cartItems[i].quantity * requiredAddons[x].quantity
            }
          }
        }
      }
    }
  }
  return { cart: count, total: count + count2 }
}

export const getCartItemQuantity = sku => {
  const cart = getCart()
  if (cart && cart.cartItems && cart.cartItems.length > 0) {
    for (let i = 0, n = cart.cartItems.length; i < n; i++) {
      if (sku === cart.cartItems[i].sku) {
        return cart.cartItems[i].quantity
      }
    }
  }
  return 0
}

export function addToCart(product, price, activeAddons) {
  clearCheckoutState()
  let cart = getCart()
  let cartItems = cart && cart.cartItems ? cart.cartItems : []
  let alreadyInCart = false
  let existingProduct
  for (let i = 0; i < cartItems.length; i++) {
    if (
      cartItems[i].sku === product.sku &&
      (JSON.stringify(activeAddons) === JSON.stringify(cartItems[i].activeAddons) ||
        (!activeAddons && !cartItems[i].activeAddons))
    ) {
      alreadyInCart = true
      existingProduct = cartItems[i]
    }
  }
  if (!alreadyInCart) {
    store.dispatch(
      addProductToCart({
        sku: product.sku,
        quantity: 1,
        product: cartProduct(product),
        price: price,
        activeAddons: activeAddons,
      })
    )
  } else {
    store.dispatch(updateCartProductQuantity(existingProduct.sku, parseInt(existingProduct.quantity) + 1))
  }
}

export function removeFromCart(product, index) {
  clearCheckoutState()
  let cart = getCart()
  if (cart && cart.cartItems) {
    let cartItems = cart.cartItems
    for (let i = 0; i < cartItems.length; i++) {
      if (product.sku === cartItems[i].sku && index === i) {
        const quantity = cartItems[i].quantity
        fetchProductOrGC(product.sku).then(data => {
          window.dataLayer.push({
            event: 'ee_remove',
            ecommerce: { remove: { products: [analyticsProduct(data, quantity)] } },
          })
        })
        store.dispatch(removeProductFromCart(product.sku))
      }
    }
    let cartTotal = getCartTotal(cart)
    announce('Item removed. New cart total: $' + cartTotal)
  }
  if (cart && cart.cartItems.length == 0) {
    try {
      taskDone(
        () => {
          const ele = document.getElementById('empty-cart-continue')
          if (ele) {
            ele.focus()
          }
        },
        500,
        'focusEmpty'
      )
    } catch (e) {}
  }
}

export function getCartTotal(cart) {
  let subtotal = 0
  if (cart && cart.cartItems) {
    if (cart && cart.cartItems) {
      let cartItems = cart.cartItems
      for (let i = 0; i < cartItems.length; i++) {
        subtotal = subtotal + cartItems[i].quantity * cartItems[i].price
        if (cartItems[i].activeAddons) {
          for (let x = 0, n = cartItems[i].activeAddons.length; x < n; x++) {
            subtotal =
              subtotal +
              productPrice(cartItems[i].activeAddons[x]) * cartItems[i].activeAddons[x].quantity * cartItems[i].quantity
          }
        }
      }
    }
  }
  return subtotal
}

export const verifyAndUpdateCart = async cartIn => {
  const cart = { ...cartIn }
  if (cart && cart.cartItems && cart.cartItems.length > 0) {
    const cartItems = cart.cartItems
    const region = getRegionZone().region
    const price_zone = getRegionZone().zone
    const priceString = region + '_' + price_zone
    for (let i = 0, n = cartItems.length; i < n; i++) {
      if (cartItems[i]) {
        const product = await fetchProductOrGC(cartItems[i].sku)
        cartItems[i].product = cartProduct(product)
        const regionSalePrice = product.pricing[priceString + '_sale_price']
        const regionListPrice = product.pricing[priceString + '_list_price']
        if (regionSalePrice) {
          cartItems[i].price = regionSalePrice
        } else if (regionListPrice) {
          cartItems[i].price = regionListPrice
        } else {
          cartItems[i].price = product.pricing.default_price
        }
        if (i === n - 1) {
          cart.cartItems = cartItems
        }
      }
    }
  }
  saveLocalStorage('cart', cart)
}

export const checkProductRegionAvailability = async regionIn => {
  const rtg_location = getCurrentLocation()
  const region = regionIn ? regionIn : rtg_location ? rtg_location.region : 'FL'
  let cart = getCart()
  if (!cart || (cart && !cart.cartItems) || (cart && cart.cartItems && cart.cartItems.length < 1)) {
    cart = getFromBrowserStorage('local', 'cart')
  }
  if (cart && cart.cartItems && region) {
    let productPromises = []
    for (let i = 0; i < cart.cartItems.length; i++) {
      productPromises.push(
        fetchProductOrGC(cart.cartItems[i].sku).then(data => {
          let newData = data
          const product = cart.cartItems[i] && cart.cartItems[i].product
          if (
            rtg_location &&
            product &&
            (rtg_location.region !== 'OOM' ||
              ((product.delivery_type === 'O' || product.delivery_type === 'U') && rtg_location.region === 'OOM')) &&
            product.category !== 'gift-card'
          ) {
            return fetchProductWarehouseAvailability(data.sku, rtg_location.distribution_index, rtg_location.state)
              .then(data => {
                const stockMessage = availabilityStockMessage(data, product, rtg_location)
                if (
                  stockMessage === 'Discontinued' ||
                  stockMessage === 'Available Soon' ||
                  stockMessage === 'Out of Stock'
                ) {
                  newData.stock = false
                } else {
                  newData.stock = true
                }
                return newData
              })
              .catch(() => {
                newData.stock = false
                return newData
              })
          } else {
            newData.stock = true
            return newData
          }
        })
      )
    }
    return Promise.all(productPromises).then(products => {
      let skusNotAvailable = []
      for (let i = 0, n = products.length; i < n; i++) {
        if (products[i].category !== 'gift-card') {
          if (!products[i].stock) {
            skusNotAvailable.push({
              sku: products[i].sku,
              reason: 'stock',
            })
          } else if (
            !products[i].catalog_availability[region] &&
            products[i].delivery_type !== 'O' &&
            products[i].delivery_type !== 'U'
          ) {
            skusNotAvailable.push({
              sku: products[i].sku,
              reason: 'region',
            })
          }
        }
      }
      return skusNotAvailable
    })
  } else {
    return []
  }
}

export const getProductsFromCart = cart => {
  const cartObj = cart || getCart()
  let products = []
  let productPromises = []
  if (cartObj && cartObj.cartItems) {
    if (cartObj.cartItems.length > 0) {
      for (let i = 0; i < cartObj.cartItems.length; i++) {
        productPromises.push(fetchProductOrGC(cartObj.cartItems[i].sku))
      }
    } else {
      for (let item in cartObj.cartItems) {
        cartObj.cartItems[item]['product']['quantity'] = cartObj.cartItems[item]['quantity']
        products.push(cartObj.cartItems[item]['product'])
      }
    }
  }
  return Promise.all(productPromises)
}

export const getRoomsToGoDeliveryItems = cartItems => {
  let deliveryItems = []
  for (let i = 0, n = cartItems.length; i < n; i++) {
    let deliveryTypeBlacklist = ['O', 'U', 'T']
    if (!deliveryTypeBlacklist.includes(cartItems[i].product.delivery_type)) {
      deliveryItems.push(cartItems[i])
    }
  }
  return deliveryItems
}

export const getOtherDeliveryItems = cartItems => {
  let deliveryItems = []
  for (let i = 0, n = cartItems.length; i < n; i++) {
    let deliveryTypeWhitelist = ['O', 'U', 'T']
    if (deliveryTypeWhitelist.includes(cartItems[i].product.delivery_type)) {
      deliveryItems.push(cartItems[i])
    }
  }
  return deliveryItems
}

export const cartProduct = (product, quantity) => {
  if (product) {
    let newProduct = product
    const loc = getCurrentLocation()
    if (product.addon_items && product.addon_items.length > 0) {
      for (let i = 0, n = product.addon_items.length; i < n; i++) {
        delete newProduct.addon_items[i]['catalog_availability']
        for (let key in product.addon_items[i].pricing) {
          if (!key.includes(`${ loc.region }_${ loc.price_zone }`) && !key.includes('default')) {
            delete newProduct.addon_items[i].pricing[key]
          }
        }
      }
    }
    if (product.items_in_room) {
      if (product.items_in_room.FL || product.items_in_room.SE || product.items_in_room.TX) {
        for (let key in product.items_in_room) {
          if (key.includes(loc.region)) {
            newProduct.items_in_room = newProduct.items_in_room[key]
          }
        }
      }
    }
    return {
      title: newProduct.title,
      sku: newProduct.sku,
      price: productPrice(newProduct),
      list_price: productPrice(newProduct, true),
      unitPrice: newProduct.unitPrice,
      delivery_type: newProduct.delivery_type,
      category: newProduct.category,
      quantity: newProduct.quantity || quantity,
      free_shipping: newProduct.free_shipping,
      on_sale: productOnSale(newProduct),
      primary_image: newProduct.primary_image,
      addon_items: newProduct.addon_items,
      package_skus: newProduct.package_skus,
      items_in_room: newProduct.items_in_room,
      shipping_cost_code: newProduct.shipping_cost_code,
      strikethrough: newProduct.strikethrough,
    }
  } else {
    return null
  }
}

export const compareLineItems = (cartLineItems, orderLineItems) => {
  for (let i = 0, n = cartLineItems.length; i < n; i++) {
    const orderItems = orderLineItems.filter(orderItem => orderItem.sku === cartLineItems[i].sku)
    let quantity = 0
    for (let x = 0, y = orderItems.length; x < y; x++) {
      quantity += orderItems[x].quantity
    }
    if (quantity !== cartLineItems[i].quantity) {
      return false
    }
  }
  return true
}

export const setNewQuantity = (event, cart, sku, activeAddons) => {
  if (cart && cart.cartItems) {
    let cartItems = cart.cartItems
    for (let i = 0; i < cartItems.length; i++) {
      const quantity = parseInt(event.target.value)
      const productQuantity = cartItems[i].quantity
      if (cartItems[i].sku === sku && JSON.stringify(activeAddons) === JSON.stringify(cartItems[i].activeAddons)) {
        fetchProductOrGC(sku).then(data => {
          if (productQuantity < quantity) {
            window.dataLayer.push({
              event: 'ee_add',
              ecommerce: { add: { products: [analyticsProduct(data, quantity - productQuantity)] } },
            })
          } else if (productQuantity > quantity) {
            window.dataLayer.push({
              event: 'ee_remove',
              ecommerce: { remove: { products: [analyticsProduct(data, productQuantity - quantity)] } },
            })
          }
        })
        store.dispatch(updateCartProductQuantity(cartItems[i].sku, event.target.value))
        addToDataLayer('click', 'cart', 'change quantity', `${ sku }, ${ event.target.value }`)
      }
    }
    announce('Quantity Updated. New cart total: $' + getCartTotal(cart))
  }
}

export const cartUpdate = async (
  cart,
  skusNotAvailableIn,
  apiCalled,
  discount,
  setCartState,
  setApiCalled,
  updateOrder = false
) => {
  const region = getRegionZone().region
  if (cart && cart.cartItems && region) {
    checkProductRegionAvailability().then(skusNotAvailable => {
      if (skusNotAvailableIn.length === 0 || skusNotAvailable.length === 0) {
        setCartState({
          skusNotAvailable: skusNotAvailable,
        })
      }
    })
    let order = getOrder()
    const location = getCurrentLocation()
    const lineItems = getLineItems()
    let verifyQuantities
    if (order && order.lineItems) {
      verifyQuantities = compareLineItems(lineItems, order.lineItems)
    }
    const orderStorage = getFromBrowserStorage('session', 'order')
    if (!orderStorage) {
      order && store.dispatch(setOrder(order))
    }
    if (
      location &&
      lineItems &&
      lineItems.length > 0 &&
      (!order.orderId || updateOrder || verifyQuantities === false)
    ) {
      if (!apiCalled) {
        setApiCalled(true)
        if (!order.orderId) {
          order = await createOrder({
            lineItems: lineItems,
            region: location.region,
            zone: parseInt(location.price_zone),
            distribution_index: parseInt(location.distribution_index),
          })
          store.dispatch(setCheckoutStep('shipping'))
        } else {
          order = await updateLineItems({
            orderId: order.orderId,
            lineItems: lineItems,
            region: location.region,
            zone: parseInt(location.price_zone),
            distribution_index: parseInt(location.distribution_index),
          })
          order.total = order.total - order.tax - order.totalDeliveryCharge
          order.totalDeliveryCharge = 0
          order.tax = 0
          store.dispatch(setCheckoutStep('shipping'))
        }
        setApiCalled(false)
      }
      order && store.dispatch(setOrder(order))
    }
    setDiscount(order, discount, setCartState)
  }
}

export const setDiscount = (order, discountIn, setCartState) => {
  let discount = 0
  if (order && order.promotions && order.promotions.totalSavings && order.promotions.totalSavings > 0) {
    discount = parseFloat(order.promotions.totalSavings)
  }
  if (discount !== discountIn) {
    setCartState({
      showPayPal: true,
      discount: discount,
    })
  } else {
    setCartState({
      showPayPal: true,
    })
  }
}

export const removeUnavailableItems = (setCartState, skusNotAvailableIn) => {
  let cart = getCart()
  if (cart && cart.cartItems) {
    for (let i = cart.cartItems.length - 1; i >= 0; i--) {
      let skusNotAvailable = []
      skusNotAvailableIn.map(avail => skusNotAvailable.push(avail.sku))
      if (skusNotAvailable.includes(cart.cartItems[i].sku)) {
        cart.cartItems.splice(i, 1)
      }
    }
    if (cart.cartItems && cart.cartItems.length === 0) {
      cart = { cartItems: [] }
    }
    for (let key in sessionStorage) {
      if (key !== 'shippingInfo' && key !== 'cart') sessionStorage.removeItem(key)
    }
    store.dispatch(setCart(cart))
    setCartState({
      skusNotAvailable: [],
    })
  }
}

export const loadCartFromOrder = async orderId => {
  fetchOrder({ orderId: orderId }).then(order => {
    let cartItems = []
    if (order && order.lineItems && order.lineItems.length > 0) {
      order.lineItems.map(item => {
        cartItems.push(
          fetchProductOrGC(item.sku).then(product => {
            return {
              sku: product.sku,
              quantity: item.quantity,
              product: cartProduct(product),
              price: productPrice(product),
              activeAddons: null,
            }
          })
        )
      })
      Promise.all(cartItems).then(cartItems => {
        store.dispatch(setCart({ cartItems: cartItems }))
      })
    }
  })
}

export const getPromoTargetSkus = () => {
  const order = getOrder()
  if (order && order.promotions && order.promotions.availableTargetSkus) {
    return order.promotions.availableTargetSkus
  }
}
