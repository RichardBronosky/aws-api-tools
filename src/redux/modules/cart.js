import { saveLocalStorage } from '../../lib/helpers/storage'

// Actions
export const SET_CART = 'SET_CART'
export const ADD_PRODUCT_TO_CART = 'ADD_PRODUCT_TO_CART'
export const UPDATE_CART_PRODUCT_QUANTITY = 'UPDATE_CART_PRODUCT_QUANTITY'
export const REMOVE_PRODUCT_FROM_CART = 'REMOVE_PRODUCT_FROM_CART'
export const CLEAR_CART = 'CLEAR_CART'
export const ADD_ACTIVE_ADDON = 'ADD_ACTIVE_ADDON'
export const REMOVE_ACTIVE_ADDON = 'REMOVE_ACTIVE_ADDON'

// Action Creators
export const setCart = cart => {
  if (cart && cart.cartItems && cart.cartItems.length > 0) {
    saveLocalStorage('cart', cart)
  }
  return { type: SET_CART, cart }
}

export const addProductToCart = product => {
  return { type: ADD_PRODUCT_TO_CART, product }
}

export const updateCartProductQuantity = (sku, quantity) => {
  return { type: UPDATE_CART_PRODUCT_QUANTITY, sku, quantity }
}

export const removeProductFromCart = sku => {
  return { type: REMOVE_PRODUCT_FROM_CART, sku }
}

export const clearCart = () => {
  return { type: CLEAR_CART }
}

export const addActiveAddon = (productSku, addon, cartIndex) => {
  return { type: ADD_ACTIVE_ADDON, productSku, addon, cartIndex }
}

export const removeActiveAddon = (productSku, addonSku, cartIndex) => {
  return { type: REMOVE_ACTIVE_ADDON, productSku, addonSku, cartIndex }
}

// Reducer
const initialState = {
  cart: {
    cartItems: [],
  },
}

export default (state = initialState, action) => {
  const newState = { ...state }
  switch (action.type) {
    case SET_CART:
      return {
        ...newState,
        cart: action.cart,
      }
    case ADD_PRODUCT_TO_CART:
      newState.cart.cartItems.push(action.product)
      saveLocalStorage('cart', newState.cart)
      return {
        ...state,
        cart: { ...newState.cart },
      }
    case UPDATE_CART_PRODUCT_QUANTITY:
      newState.cart.cartItems.map(
        obj =>
          (newState.cart.cartItems.find(product => {
            return product.sku === action.sku
          }).quantity = parseInt(action.quantity) || obj)
      )
      saveLocalStorage('cart', newState.cart)
      return {
        ...state,
        cart: { ...newState.cart },
      }
    case REMOVE_PRODUCT_FROM_CART:
      const index = newState.cart.cartItems.findIndex(product => {
        return product.sku === action.sku
      })
      newState.cart.cartItems.splice(index, 1)
      saveLocalStorage('cart', newState.cart)
      return {
        ...state,
        cart: { ...newState.cart },
      }
    case CLEAR_CART:
      saveLocalStorage('cart', { cartItems: [] })
      return {
        ...state,
        cart: { cartItems: [] },
      }
    case ADD_ACTIVE_ADDON: {
      const product = newState.cart.cartItems[action.cartIndex]
      product.activeAddons = product.activeAddons ? product.activeAddons : []
      product.activeAddons.push(action.addon)
      newState.cart.cartItems[action.cartIndex] = product
      saveLocalStorage('cart', newState.cart)
      return {
        ...state,
        cart: { ...newState.cart },
      }
    }
    case REMOVE_ACTIVE_ADDON: {
      const product = newState.cart.cartItems[action.cartIndex]
      const addonIndex = product.activeAddons.findIndex(addon => addon.sku === action.addonSku)
      product.activeAddons.splice(addonIndex, 1)
      product.activeAddons = product.activeAddons.length === 0 ? null : product.activeAddons
      newState.cart.cartItems[action.cartIndex] = product
      saveLocalStorage('cart', newState.cart)
      return {
        ...state,
        cart: { ...newState.cart },
      }
    }
    default:
      return state
  }
}
