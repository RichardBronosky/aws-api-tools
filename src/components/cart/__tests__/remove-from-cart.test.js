import React from 'react'
import renderer from 'react-test-renderer'
import RemoveFromCart from '../cart-parts/remove-from-cart'
import * as cartHelper from '../../../lib/helpers/cart'
import { mockProduct1 } from '../../../lib/mocks/productDataMocks'

describe('RemoveFromCart', () => {
  const testProps = {
    product: mockProduct1,
    index: 1,
  }

  it('renders correctly', () => {
    const tree = renderer.create(<RemoveFromCart { ...testProps } />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when in header', () => {
    const tree = renderer.create(<RemoveFromCart { ...testProps } headerCart />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('run removeFromCart when remove from cart clicked', () => {
    const tree = renderer.create(<RemoveFromCart { ...testProps } />)
    cartHelper.removeFromCart = jest.fn()
    tree.root.findByProps({ className: 'remove-btn' }).props.onClick()
    expect(cartHelper.removeFromCart).toHaveBeenCalled()
  })
})
