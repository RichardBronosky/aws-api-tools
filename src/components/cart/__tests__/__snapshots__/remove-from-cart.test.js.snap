// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveFromCart renders correctly 1`] = `
<button
  className="remove-btn"
  onClick={[Function]}
  value="Remove From Cart"
>
  remove 
  <span
    className="hide508"
  >
    
     from cart
  </span>
</button>
`;

exports[`RemoveFromCart renders correctly when in header 1`] = `
<button
  className="close"
  onClick={[Function]}
  value="Remove From Cart"
>
  
  <span
    className="hide508"
  >
    remove
     from cart
  </span>
</button>
`;
